### PR TITLE
Feature: wrapper for callbacks

### PR DESCRIPTION
**Pull request by ditwoo**
## Before submitting

- [x] Was this discussed/approved via a Github issue? (no need for typos and docs improvements)
- [x] Did you read the [contribution guide](https://github.com/catalyst-team/catalyst/blob/master/CONTRIBUTING.md)?
- [x] Did you check the code style? `catalyst-make-codestyle && catalyst-check-codestyle` (`pip install -U catalyst-codestyle`).
- [x] Did you make sure to update the docs? We use Google format for all the methods and classes.
- [x] Did you check the docs with `make check-docs`?
- [x] Did you write any new necessary tests?
- [x] Did you add your new functionality to the docs?
- [x] Did you update the [CHANGELOG](https://github.com/catalyst-team/catalyst/CHANGELOG.md)?
- [x] **You can use 'Login as guest' to see Teamcity build logs.**

## Type of Change

- [x] Examples / docs / tutorials / contributors update
- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] Improvement (non-breaking change which improves an existing feature)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## PR review

Anyone in the community is free to review the PR once the tests have passed.
If we didn't discuss your PR in Github issues there's a high chance it will not be merged.

### CODE REVIEW

File: catalyst/contrib/dl/callbacks/ignore_metric_callback.py [hunk 1]
```diff
@@ -0,0 +1,78 @@
+from catalyst.core.callback import Callback, CallbackOrder
+from catalyst.core.runner import IRunner
+
+
+class IgnoreMetricCallback(Callback):
```
Comment (Scitator) [inline catalyst/contrib/dl/callbacks/ignore_metric_callback.py]: just to be sure, could you please update docs with this new callbacks & changelog :)
I now it only draft yet :)

Comment (Scitator) [reply catalyst/contrib/dl/callbacks/ignore_metric_callback.py]: speaking about architecture and implementation,
what do you think about `CallbackFilterCallback`?
not only with name filtering, but also with some condition one?
```python
callback_filter = CallbackFilterCallback(
  key={"train": "metric-1"}, 
  lambda={"valid": lambda key, value: isinstance(value, MetricCallback)},
  # or, even we can also add something like
  value={"valid": MetricCallback}
)
```

long story shot, with callback filtering, most of the time you have 2 options:
1. filter by name (`key` field)
2. filter by class (`value` field)
3. some very custom condition -> lambda

what do you think about such proposal?

Comment (Scitator) [reply catalyst/contrib/dl/callbacks/ignore_metric_callback.py]: moreover, if we would like to think about filtering in a broader sense....
why do we filter only on loader name? during stage, we have 2 fors - epoch and loader one.
In this case, I think, we could filter not only on loader name, but also on epoch index.

nevertheless, still now sure about user friendly API for this :)

Comment (Scitator) [reply catalyst/contrib/dl/callbacks/ignore_metric_callback.py]: maybe, only maybe, something like

```python
callback_filter = CallbackFilterCallback(
  condition=lambda epoch, loader: loader == "train",
  filter=lambda key, value: isinstance(value, MetricCallback),
)
```

Comment (ditwoo) [reply catalyst/contrib/dl/callbacks/ignore_metric_callback.py]: @Scitator what do you think about `FilterCallback` or `SkipCallback` or `FilterMetricCallback` as a name for a class?
(I think that we can generate a better name without repeated *callback* word)

Also what do you think about use cases for arguments:

```python
callback = FilterCallback(
    loader0="to_filter",  # base case - ignore callback for loader
    loader1=["to_fitler1", "to_filter2"],  # ignore multiple callbacks
    loader2={"to_filter1": [1, 22, 333], "to_filter2": [1, 2, 3]},  # turn off callbacks for epochs
    loader3=lambda epoch, loader_name, loader_obj: loader_obj,  # filter with custom function
)
```

Comment (Scitator) [reply catalyst/contrib/dl/callbacks/ignore_metric_callback.py]: @Ditwoo I think, `FilterCallback` would be good choice :)
First of all, we should not concentrate on Metric-based callbacks only.

File: catalyst/dl/experiment/config.py [hunk 2]
```diff
@@ -430,6 +431,23 @@ def _get_callback(**params):
             return ConfigExperiment._get_callback(**wrapper_params)
         return callback
 
+    @staticmethod
+    def _is_same_types(first: Callback, second: Callback) -> bool:
```
Comment (Scitator) [inline catalyst/dl/experiment/config.py]: let's move this function to 
```
catalyst/dl/utils/callbacks.py
```
so it would be easier to use it from anywhere

Comment (Scitator) [reply catalyst/dl/experiment/config.py]: and what do you think about splitting it into two separate callbacks, like
```python
def get_original_callback(callback) -> callback:
  while not callback and is-wrapper:
    return callback.callback  # unwrap

def check_isinstance(first, second) -> bool:
  first = get_original_callback(first)
  second = get_original_callback(second)
  output = isinstance(first, second)
  return output
```

File: bin/tests/check_dl_core_wrapper_callback.sh [hunk 3]
```diff
@@ -0,0 +1,265 @@
...
+LOG_MSG='pipeline 00'
+echo ${LOG_MSG}
+
+LOGDIR=./tests/logs/_tests_dl_callbacks
+CHECKPOINTS=${LOGDIR}/checkpoints
+LOGFILE=${CHECKPOINTS}/_metrics.json
+EXP_OUTPUT=./tests/output.txt
+
+PYTHONPATH=./examples:./catalyst:${PYTHONPATH} \
+  python3 -c "
```
Comment (Scitator) [inline bin/tests/check_dl_core_wrapper_callback.sh:72]: btw, could you please move your tests to .py files,
like https://github.com/catalyst-team/catalyst/blob/master/catalyst/contrib/dl/callbacks/tests/test_tracer_callback.py
?

File: catalyst/contrib/dl/callbacks/periodic_loader_callback.py [hunk 4]
```diff
@@ -12,13 +12,31 @@ class PeriodicLoaderCallback(Callback):
...
+    3 epochs and ``valid_additional`` - every 5 epochs:
+
+    .. code-block:: python
+
+        from catalyst.dl import (
+            SupervisedRunner, PeriodicLoaderRunnerCallback,
+        )
+        runner = SupervisedRunner()
+        runner.train(
+            ...
```
Comment (Scitator) [inline catalyst/contrib/dl/callbacks/periodic_loader_callback.py:27]: would you like to add `loaders={"train_additional": ..., "valid": ..., "valid_additional":...}`?

File: catalyst/core/callbacks/wrapper.py [hunk 5]
```diff
@@ -0,0 +1,298 @@
...
+    need to wrap callback:
+
+    .. code-block:: python
+
+        from catalyst.dl import (
+            SupervisedRunner, WrapperCallback, CriterionCallback
+        )
+        runner = SupervisedRunner()
+        runner.train(
+            ...
```
Comment (Scitator) [inline catalyst/core/callbacks/wrapper.py]: 🤔  what do you think about full-featured examples?
we have really tiny one:
```python
import torch
from torch.utils.data import DataLoader, TensorDataset
from catalyst.dl import SupervisedRunner

num_samples, num_features = int(1e4), int(1e1)
X, y = torch.rand(num_samples, num_features), torch.rand(num_samples)
loader = DataLoader(TensorDataset(X, y), batch_size=32, num_workers=1)
loaders = {"train": loader, "valid": loader}

model = torch.nn.Linear(num_features, 1)
criterion = torch.nn.MSELoss()
optimizer = torch.optim.Adam(model.parameters())
scheduler = torch.optim.lr_scheduler.MultiStepLR(optimizer, [3, 6])

runner = SupervisedRunner()
runner.train(
    model=model, criterion=criterion, optimizer=optimizer, scheduler=scheduler,
    loaders=loaders, logdir="./logdir", num_epochs=8, verbose=True,
)
```

File: catalyst/core/callbacks/wrapper.py [hunk 6]
```diff
@@ -0,0 +1,298 @@
...
+            callback: CriterionCallback
+          ...
+
+    """
+
+    def __init__(
+        self,
+        base_callback: Callback,
+        loaders: LOADERS = None,
+        filter_fn: Union[str, FILTER_FN] = None,
```
Comment (Scitator) [inline catalyst/core/callbacks/wrapper.py:229]: would you like to add `epoch` support to run callback every `epoch` epochs? 😂

File: catalyst/dl/utils/callbacks.py [hunk 7]
```diff
@@ -0,0 +1,33 @@
...
+
+    Returns:
+        callback inside wrapper
+    """
+    while isinstance(callback, WrapperCallback):
+        callback = callback.callback
+    return callback
+
+
+def check_isinstance(first: Callback, second: Callback) -> bool:
```
Comment (Scitator) [inline catalyst/dl/utils/callbacks.py]: `check_callback_isinstance`?

File: docs/api/core.rst [hunk 8]
```diff
@@ -129,6 +129,12 @@ Validation
     :undoc-members:
     :show-inheritance:
 
+Wrapper
+~~~~~~~~~~~~~~~~~~~~~~
+.. automodule:: catalyst.core.callbacks.wrapper
```
Comment (Scitator) [inline docs/api/core.rst:134]: ❤️

File: catalyst/contrib/dl/callbacks/periodic_loader_callback.py (hunk 4 above)
Comment (ditwoo) [reply catalyst/contrib/dl/callbacks/periodic_loader_callback.py:27]: yep, i agree, need to add this

File: catalyst/core/callbacks/wrapper.py (hunk 6 above)
Comment (ditwoo) [reply catalyst/core/callbacks/wrapper.py:229]: sure

File: catalyst/core/callbacks/wrapper.py [hunk 9]
```diff
@@ -0,0 +1,361 @@
...
+            optimizer=optimizer,
+            scheduler=scheduler,
+            loaders=loaders,
+            logdir="./logdir",
+            num_epochs=8,
+            verbose=True,
+            callbacks=[
+                WrapperCallback(
+                    base_callback=CriterionCallback(),
+                    loaders=["valid"]
```
Comment (Scitator) [inline catalyst/core/callbacks/wrapper.py]: 🤔  based on provided example, should we split it's functionality to 
```
loaders=[] # loader names, where callback should be executed
ignore_loaders=[] # loader names, where we should ignore the callback
```
?

just wondering, how it would be more user-friendly

Comment (Scitator) [reply catalyst/core/callbacks/wrapper.py]: btw, it also looks like this examples should not work because by default our metric - "loss" on "valid" loader, that is not computed here :)

Comment (Scitator) [reply catalyst/core/callbacks/wrapper.py]: I think, we could easily use `dl.AccuracyCallback` instead

File: catalyst/core/callbacks/wrapper.py [hunk 10]
```diff
@@ -289,6 +354,9 @@ def on_loader_start(self, runner: IRunner) -> None:
         if self.filter_fn is not None:
             self._is_active = not self.filter_fn(stage, epoch, loader)
 
+        # import inspect
+        # print(">>>", inspect.getsource(self.filter_fn))
```
Comment (Scitator) [inline catalyst/core/callbacks/wrapper.py]: not sure if we need this :)

File: catalyst/core/callbacks/wrapper.py [hunk 11]
```diff
@@ -29,18 +31,28 @@ def _filter_fn_from_epochs(
 
     if isinstance(epochs, (list, tuple)):
         epochs = sorted(set(epochs))
-        filter_fn = lambda stage, epoch, loader: epoch in epochs
+        if reverse_condition:
+            filter_fn = lambda stage, epoch, loader: epoch not in epochs
+        else:
+            filter_fn = lambda stage, epoch, loader: epoch in epochs
```
Comment (Scitator) [inline catalyst/core/callbacks/wrapper.py:40]: do you think epoch list would be user friendly?
from my point of view most of the time you want to execute something `one time per N epochs`

maybe we should change the above condition to `epoch % specified_epoch == 0`?

File: catalyst/core/callbacks/wrapper.py (hunk 10 above)
Comment (ditwoo) [reply catalyst/core/callbacks/wrapper.py]: it was for debug 😃

File: catalyst/core/callbacks/wrapper.py (hunk 11 above)
Comment (ditwoo) [reply catalyst/core/callbacks/wrapper.py:40]: @Scitator not everywhere.
i think we should keep both:
- if specified number (for example `N`) then disable/enable callback every `N` epochs
- if specified list of numbers then disable/enable callback at specified epochs

also, if we touch `epochs` did we expect similar logic from `loaders` argument?
in current implementation it supports only list of epochs.
do we need the same logic for `loaders`:

```
loaders={
     <loader 1>: <number>,  # should we do some periodical action for this loader?
     <loader 2>: [<number 1>, <number 2>, <number 3>],  # disable/enable callback for specified epochs
}
```

?

Comment (Scitator) [reply catalyst/core/callbacks/wrapper.py:40]: 🤔  I think, you are right about 2 cases support for epochs
nevertheless, for loaders I suppose we only have enable/disable action

File: catalyst/core/callbacks/wrapper.py [hunk 12]
```diff
@@ -8,6 +8,113 @@
 FILTER_FN = Callable[[str, int, str], bool]
 
 
+class WrapperCallback(Callback):
```
Comment (Scitator) [inline catalyst/core/callbacks/wrapper.py]: maybe,  we should move it to `catalyst/core/callback`?

Comment (ditwoo) [reply catalyst/core/callbacks/wrapper.py]: sure

**Review by Scitator — APPROVED** [review]
